Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -0,0 +1,334 @@
+      * Sets the current value, iff a model is associated with the value.
+      * @param value The current value.
+      * @return Acceptable if the value was set successfully, Invalid
+      *         otherwise.
+      */
+      QValidator::State set_current(const T& value);
+
+      /** Connects a slot to the current signal. */
+      boost::signals2::connection connect_current_signal(
+        const typename CurrentSignal::slot_type& slot) const;
+
+    private:
+      mutable CurrentSignal m_current_signal;
+      T m_current;
+      std::unordered_map<T, Details::AssociatedBooleanModel> m_models;
+      bool m_is_blocked;
+
+      void set_associated_model_value(const T& value, bool model_value);
+      void on_current(const T& value, bool is_selected);
+  };
+
+  /**
+  * A specialized AssociativeValueModel, allowing all associated models to have
+  * a value of false.
+  */
+  template <typename T>
+  class AssociativeValueModel<boost::optional<T>> :
+      public ValueModel<boost::optional<T>> {
+    public:
+
+      AssociativeValueModel();
+
+      void associate(const std::shared_ptr<BooleanModel>& model,
+        const T& value);
+
+      std::shared_ptr<BooleanModel> make_association(const T& value);
+
+      std::shared_ptr<BooleanModel> find(const T& value) const;
+
+      QValidator::State get_state() const;
+
+      /**
+      * Returns the current value. The returned value is none if no
+      * associated model has a value of true.
+      * @return The current value.
+      */
+      const boost::optional<T>& get_current() const;
+
+      /**
+      * Sets the current value if the value has an associated model. If the
+      * given value is none, any associated model that has a value of true
+      * is set to false.
+      * @param value The current value.
+      * @return Acceptable if the value was set successfully, Invalid
+      *         otherwise.
+      */
+      QValidator::State set_current(const boost::optional<T>& value);
+
+      boost::signals2::connection connect_current_signal(
+        const typename ValueModel<boost::optional<T>>::CurrentSignal::slot_type& slot) const;
+
+    private:
+      mutable typename ValueModel<boost::optional<T>>::CurrentSignal
+        m_current_signal;
+      boost::optional<T> m_current;
+      std::unordered_map<T, Details::AssociatedBooleanModel> m_models;
+      bool m_is_blocked;
+
+      void set_associated_model_value(const T& value, bool model_value);
+      void on_current(const T& value, bool is_selected);
+  };
+
+  template <typename T>
+  AssociativeValueModel<T>::AssociativeValueModel()
+    : m_is_blocked(false) {}
+
+  template <typename T>
+  void AssociativeValueModel<T>::associate(
+      const std::shared_ptr<BooleanModel>& model, const T& value) {
+    if(m_models.find(value) != m_models.end()) {
+      return;
+    }
+    if(model->get_current() == true) {
+      model->set_current(false);
+    }
+    m_models[value] = {model, model->connect_current_signal(
```

Comment:
Don't insert values using this approach, the ```operator []``` for updating values, not inserting new ones.